guides/javascript.md

```diff
@@ -0,0 +1,215 @@
+
+If the name represents a negative statement, reword it into a positive statement and inverting the value assigned to the value.
+
+🚫
+
+``` javascript
+const notEnoughGas = false;
+```
+
+✅
+
+``` javascript
+const hasEnoughGas = true;
+```
+
+Take special note of variables which are created via React's `useState` hook.
+
+🚫
+
+``` javascript
+const [removed, setRemoved] = useState(false);
+```
+
+✅
+
+``` javascript
+const [isRemoved, setIsRemoved] = useState(false);
+```
+
+## 💡 Place names of secondary concepts first in boolean variable names
+
+Sometimes a boolean variable describes a subject that's different than the context where the variable is defined:
+
+``` javascript
+const ownedAccount = Boolean(ownedAccountName);
+```
+
+In naming such a variable, there are a couple of different routes you could take.
+
+One way is to always place the verb at the beginning:
+
+``` javascript
+const isRecipientOwnedAccount = Boolean(ownedAccountName);
+```
+
+However, this naming strategy creates a point of friction for objects, arrays, or React components, where it may be desirable to sort identifiers alphabetically. In that case you could end up with something like:
+
+``` jsx
+<SenderToRecipient
+  isRecipientOwnedAccount={isRecipientOwnedAccount}
+  recipientName={toName}
+  recipientNickname={toNickname}
+  senderName={fromName}
+  senderAddress={fromAddress}
+/>
+```
+
+It is potentially more "greppable" if properties that concern the same concept are kept together instead of separate. To address this, you may wish to place the subject of the variable name at the beginning:
+
+``` javascript
+const recipientIsOwnedAccount = Boolean(ownedAccountName);
+```
+
+This would result in:
+
+``` jsx
+<SenderToRecipient
+  recipientIsOwnedAccount={isRecipientOwnedAccount}
+  recipientName={toName}
+  recipientNickname={toNickname}
+  senderName={fromName}
+  senderAddress={fromAddress}
+/>
+```
+
+## Use `async`/`await` syntax over `.then`/`.catch`
+
+Asynchronous code written using `async`/`await` syntax looks less complex and more straightforward than code written using `.then`/`.catch`. Additionally, using `async`/`await` leads to [better stack traces in Node and Chromium](https://mathiasbynens.be/notes/async-stack-traces), both of which use the V8 JavaScript engine, because when an asynchronous operation is `await`ed, the engine will remember the function where the `await` occurred, which means that it can place that function on the stack trace (otherwise, using `.then`/`.catch`, it would get lost).
+
+🚫
+
+``` javascript
+function makeRequest() {
+  return fetch("https://google.com")
+    .then((response) => {
+      return response.json().then((json) => {
+        return json['page_views'];
+      })
+    })
+}
+```
+
+✅
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch("https://google.com");
+  const json = await response.json();
+  return json['page_views'];
+}
+```
+
+## `await` promises before `return`ing them
```

Comment:
I agree with this, but it should be noted that this makes code significantly slower:

https://arthur.place/the-cost-of-return-await

Reply:
Good point! I can add a disclaimer.

Reply:
It looks like this can also be enforced with the [`return-await`](https://typescript-eslint.io/rules/return-await/) rule.

Reply:
By the way, for posterity, I read the article you linked to and came to [this bit at the end](https://arthur.place/the-cost-of-return-await#zero-cost-async-stack-traces) which suggests that the benefit we think `await` is providing here may not be necessary, but I don't think that's true. I am going to add an example to this section to prove that `await` is still necessary to preserve async functions in backtraces. I think [zero-cost async stack traces](https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit) is the thing that makes this even possible, but [that was added way back in v8 7.3](https://v8.dev/blog/v8-release-73#async-stack-traces) (or [Node 12, which used v7 7.8](https://nodejs.org/en/download/releases)), so I don't think we need to even mention that.

Reply:
Here's the example I used last time in a Slack discussion about this: https://consensys.slack.com/archives/G1L7H42BT/p1646953758233329?thread_ts=1646952599.353159&cid=G1L7H42BT

Reply:
I made some updates here: https://github.com/MetaMask/contributor-docs/compare/acbcc47...d632a27

Reply:
Hah, I just noticed that the referenced negative consequences don't apply to us anyway. We haven't transpiled async/await [since mid-2019](https://github.com/MetaMask/metamask-extension/pull/6812). The benchmarks shown for native async/await look pretty decent.

Reply:
Okay, so it sounds like we want to remove this disclaimer then?

Reply:
Never mind, I see both thumbs-ups. Disclaimer removed here: [22a9daa](https://github.com/MetaMask/contributor-docs/pull/11/commits/22a9daa2143c7632f13018ae5431ac950357af66)